### ISSUE TITLE
Extra regex space in __prefix_line

### ISSUE THREAD

**Issue by chr-x** [fail2ban/fail2ban#1405]
### Environment:

- Fail2Ban: 0.8.13-1
- OS: Debian   8.4
- firehol  1.297-1
- [x] Fail2Ban installed via OS/distribution mechanisms
- [x] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:

Firehol delivers lines to syslog that cannot be matched by regex's in a filter.

#### Steps to reproduce

Filter, ./fh.local:
```
[INCLUDES]
before = /etc/fail2ban/filter.d/common.conf

[Definition]
failregex = ^%(__prefix_line)sIN-world:IN=eth0 .* SRC=<HOST> .*$
```
run:
```
$ fail2ban-regex \
  'Apr 19 09:02:51 thismachine kernel: [  970.699396] IN-world:IN=eth0 OUT= MAC= SRC=123.123.1.2 DST=192.168.1.31 LEN=248 TOS=0x00 PREC=0x00 TTL=64 ID=17598 DF PROTO=UDP SPT=138 DPT=138 LEN=228' \
  ./fh.local

Running tests
=============

Use   failregex file : ./fh.local
Use      single line : Apr 19 09:02:51 thismachine kernel: [  970.699396] IN-worl...


Results
=======

Failregex: 0 total

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [1] MONTH Day Hour:Minute:Second
`-

Lines: 1 lines, 0 ignored, 0 matched, 1 missed
```

Firehol's log strings would match if they had a second space in front of
'IN-world'.

Example:
```
$ fail2ban-regex \
  'Apr 19 09:02:51 thismachine kernel: [  970.699396]  IN-world:IN=eth0 OUT= MAC= SRC=123.123.1.2 DST=192.168.1.31 LEN=248 TOS=0x00 PREC=0x00 TTL=64 ID=17598 DF PROTO=UDP SPT=138 DPT=138 LEN=228' \
  ./fh.local

Running tests
=============

Use   failregex file : ./fh.local
Use      single line : Apr 19 09:02:51 thismachine kernel: [  970.699396]  IN-wor...


Results
=======

Failregex: 1 total
|-  #) [# of hits] regular expression
|   1) [1] ^\s*(<[^.]+\.[^.]+>)?\s*(?:\S+ )?(?:kernel: \[ *\d+\.\d+\] )?(?:@vserver_\S+ )?(?:(?:\[\d+\])?:\s+[\[\(]?\S*(?:\(\S+\))?[\]\)]?:?|[\[\(]?\S*(?:\(\S+\))?[\]\)]?:?(?:\[\d+\])?:?)?\s(?:\[ID \d+ \S+\])?\s*IN-world:IN=eth0 .* SRC=<HOST> .*$
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [1] MONTH Day Hour:Minute:Second
`-

Lines: 1 lines, 0 ignored, 1 matched, 0 missed
```

#### Expected behavior

Log strings of firehol can be matched out of the box.

#### Observed behavior

Log strings of firehol cannot be matched.

#### Any additional information

My findings:
I guess this issue does affect recent versions of fail2ban.

Line 54 of config/filter.d/common.conf (current master):
```
__prefix_line = \s*%(__bsd_syslog_verbose)s?\s*(?:%(__hostname)s )?(?:%(__kernel_prefix)s )?(?:@vserver_\S+ )?%(__daemon_combs_re)s?\s%(__daemon_extra_re)s?\s*
```

introduces an extra space at column 133-134 '\s'. The space is in between '__daemon_combs_re' and '__daemon_extra_re'. Both strings do not match for firehol log strings. But the space remains. 

This extra space has been introduced on 29 Apr 2013 by commit d2a953756802bd7cf63f5f5f792371f52f5cba8c.

I suggest commit bc2e89c026732eb39f19404ae5c91345c80bbfbf. Does this commit break existing filters?

#### Relevant lines from monitored log files in question:

```
Apr 19 09:02:51 thismachine kernel: [  970.699396] IN-world:IN=eth0 OUT= MAC= SRC=123.123.1.2 DST=192.168.12.12 LEN=248 TOS=0x00 PREC=0x00 TTL=64 ID=17598 DF PROTO=UDP SPT=138 DPT=138 LEN=228
```

**Comment by yarikoptic** [comment]
interesting... I think that it just that __daemon_extra_re should have that `\s` as part of its regex group, which is then allowed to be optional.  Care to furnish the PR with a sample log line from above added to sample test logs?

**Comment by sebres** [comment]
d2a953756802bd7cf63f5f5f792371f52f5cba8c has definitively a bug, it should be something like:
```diff
-__daemon_extra_re = (?:\[ID \d+ \S+\])
+__daemon_extra_re = (?:\s\[ID \d+ \S+\])
...
-__prefix_line = \s*%(__bsd_syslog_verbose)s?\s*(?:%(__hostname)s )?(?:%(__kernel_prefix)s )?(?:@vserver_\S+ )?%(__daemon_combs_re)s?\s%(__daemon_extra_re)s?\s*
+__prefix_line = \s*%(__bsd_syslog_verbose)s?\s*(?:%(__hostname)s )?(?:%(__kernel_prefix)s )?(?:@vserver_\S+ )?%(__daemon_combs_re)s?%(__daemon_extra_re)s?\s*
```

But we have in config/filter.d/common.conf many evil constellations, because by some optional entries sometimes `\s` located left (ahead), and sometimes `\s` right. The mix (because all is optional) is undefined.

I'll normalize it now.

**Comment by chr-x** [comment]
Yes, the syntax is not fully consistent. I think it is mostly about tag separation in __prefix_line. All other variables do not have leading nor trailing spaces and I suggest to keep it like that. The major issue is about tag-separation that has not been bound to the leading tag. In the same reading __daemon_extra_re should be bound to __daemon_combs_re, I guess. __daemon_extra_re only exists with a leading __daemon_combs_re.

Suggestion:
```
...
__daemon_tag = %(__daemon_combs_re)s\s%(__daemon_extra_re)s?
__prefix_line = \s*(?:%(__bsd_syslog_verbose)s\s+)?(?:%(__hostname)s )?(?:%(__kernel_prefix)s )?(?:@vserver_\S+ )?(?:%(__daemon_tag)s\s)?\s*
...
```

Maybe `'\s'` can be replaced by `' '` ? I do not find `'<tab>'` in leading columns of my log files. On the other hand, never change a winning team.

By the way, how do you colorize the diff style green and red in a comment?

**Comment by sebres** [comment]
Already fixed in #1421

> colorize the diff

\`\`\`diff
\- enter diff here ...
\+ enter diff here ...
\`\`\`

**Comment by chr-x** [comment]
Thanks for the diff style!

Sorry, when I published my last comment, I did not see you already fixed the issue.
Just for completeness, another test for common.conf, because it is there. The code stays under /tests/ (commit 4664e0dbf0b5801ce1922d1322617d8ed2930dd5).

Maybe one should use `\s+` not `\s*` for tag separation:
```diff
-__prefix_line = \s*(?:(?:%(__bsd_syslog_verbose)s)\s*)?(?:(?:%(__hostname)s)\s*)?(?:(?:%(__kernel_prefix)s)\s*)?(?:(?:@vserver_\S+)\s*)?(?:(?:%(__daemon_combs_re)s)\s*)?(?:(?:%(__daemon_extra_re)s)\s*)?
+__daemon_tag = %(__daemon_combs_re)s\s%(__daemon_extra_re)s?
+__prefix_line = \s*(?:%(__bsd_syslog_verbose)s\s+)?(?:%(__hostname)s\s+)?(?:%(__kernel_prefix)s\s+)?(?:@vserver_\S+\s+)?(?:%(__daemon_tag)s\s+)?\s*
```

**Comment by sebres** [comment]
I'm against `\s+`:
- not backwards compatible (Apart from the fact that all the parameters can be rewritten in .local and own jail's)
- `\s*` is more flexible
- currently I can't imagine the situation where it may be bad if two parameters will be found together (without space); If you have an example - give it to me :)

**Comment by yarikoptic** [comment]
hm, anal me would prefer to have explicit space if there is one to expect!  aren't they all expected within each group? so why not just `\s` and be done (without * or +)?  that is how they were pretty much, just with explicit ' '

**Comment by chr-x** [comment]
Me too I think, a space should be matched where it is expected. If not, there are additional unknown regex combination, which is not the intention. I opt for simple tag separation without '*' nor '+'. And to be fully consequent when rewriting more than the issue itself, I'd to use simple spaces ' ' not '\s', because I do not expect `<tab>`'s in the prefix part.

In all the regex'es there is one sequence I do not understand. How does '@vserver_' work? Is '@vserver_' a literal string or is it some python code?

@sebres, what do you mean with 'not backwards compatible'. Please, can you give an example.

**Comment by sebres** [comment]
Well, the fix I've made is minimalist invasive, so totally backwards compatible - all the parameters including `__prefix_line` in common can be overridden from user, in custom filter, etc.
So if the space was previously optional, and will be mandatory, then possible that some filters or jails could be no longer operational by end-user.
With other words, I know not all constellations to say some space can be made mandatory now.

In addition, all our `failregex` are built as exact as possible, most of the time even ahead (or both sides) anchored. So there is at all not necessary to made it mandatory - each space that should be found, will be always found. 
Apart from the fact, that the spaces did not wear any connotation at all.

Otherwise the fix from #1421 will be to controversial extension (any case to behold no more as a simple fix).

Therefore I'm definitively against as well as `\s+` or `\s`!
